examples/vision/supervised-contrastive-learning.py

```diff
@@ -0,0 +1,206 @@
+"""
+Title: Supervised Contrastive Learning
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2020/11/01
+Last modified: 2020/11/01
+Description: Using supervised contrastive learning for image classification.
+
+
+## Introduction
+
+[Supervised Contrastive Learning](https://arxiv.org/abs/2004.11362)
+(Prannay Khosla et al.) is a training methodology that outperforms
+ cross-entropy on supervised learning tasks.
+
+Essentially, training an image classification model with Supervised Contrastive
+Learning is performed in two phases:
+
+1. Pre-training an encoder to generate feature vectors for input images such
+that feature vectors of images in the same class will be more similar compared
+feature vectors of images in other classes.
+2. Training a classifier on top of the frozen encoder.
+
+## Setup
+"""
+
+"""shell
+pip install -q tensorflow-addons
+"""
+
+!pip install -q tensorflow-addons
+
+import tensorflow as tf
+import tensorflow_addons as tfa
+import numpy as np
+import keras 
+from keras import layers
+from keras.preprocessing.image import ImageDataGenerator
+
+"""## Prepare the data"""
+
+num_classes = 10
+input_shape = (32, 32, 3)
+
+# Load the train and test data splits
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+
+# Display shapes of train and test datasets
+print(f'x_train shape: {x_train.shape} - y_train shape: {y_train.shape}')
+print(f'x_test shape: {x_test.shape} - y_test shape: {y_test.shape}')
+
+"""## Using image data augmentation"""
+
+data_generator = ImageDataGenerator(
+  featurewise_center=True,
+  featurewise_std_normalization=True,
+  rotation_range=20,
+  horizontal_flip=True,
+  width_shift_range=0.2,
+  height_shift_range=0.2
+)
+
+data_generator.fit(x_train)
+
+"""## Build the encoder model
+
+The encoder model takes the image as an input and produce a 128-dimension
+feature vector.
+"""
+
+def create_encoder():
+  densenet = keras.applications.ResNet50V2(
+      include_top=False,
+      weights=None,
+      input_shape=input_shape,
+      pooling='avg')
+
+  inputs = keras.Input(shape=input_shape)
+  outputs = densenet(inputs)
+  model = keras.Model(inputs=inputs, outputs=outputs, name="cifar10-encoder")
+  return model
+
+encoder = create_encoder()
+encoder.summary()
+
+learning_rate = 0.001
+batch_size = 265
+hidden_units = 512
+projection_units = 128
+num_epochs = 50
+dropout_rate = 0.5
+temperature = 0.05
+
+"""## Build the classification model
+
+The classification model adds a fully-connected layer on top of the encoder,
+plus a softmax layer with the target classes.
+"""
+
+def create_classifier(encoder, trainable=True):
+
+  for layer in encoder.layers:
+    layer.trainable = trainable
+
+  inputs = keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  features = layers.BatchNormalization()(features)
+  features = layers.Dropout(dropout_rate)(features)
+  features = layers.Dense(hidden_units, 'relu')(features)
```

Comment:
You should not, in general, do relu before BN. You should do it after.

Reply:
Done